### DISCUSSION TITLE
How should I go about getting CASCADE deletes to work?

### DISCUSSION THREAD

**Discussion by ghost** [prisma/prisma#4601] [Q&A] [answered]
If I have a `Post` and it has 30 `Comment`s attached to it. It would make sense for those comments to also be deleted when the `Post` is deleted. 

Looking through the docs it looks like you have to set it up in SQL and introspect to generate the `schema.prisma` file.
Adding this line to get CASCADE deletes to work.

```sql
CONSTRAINT post FOREIGN KEY ("post_id") REFERENCES "public"."posts" (id) ON DELETE CASCADE
```

But I'm not all too familiar with SQL so I always opt to defining my models in the `schema.prisma` file and migrating from there.
Looking at the SQL generate from this approach. I found the following SQL statement in the migration folder.

```sql
ALTER TABLE "comments" ADD FOREIGN KEY("post_id")REFERENCES "posts"("id") ON DELETE CASCADE ON UPDATE CASCADE;
```

Seeing the above I figured CASCADE was setup so I tried the following code.

```typescript
t.nonNull.field('deletePost', {
  type: 'Boolean',
  args: {
    postId: nonNull(intArg())
  },
  resolve: async (_root, args, context) => {
    await context.prisma.post.delete({
      where: { id:  args.postId },
    })

    return true
  },
})
```

But I ended up getting the following error.

```
"message": "\nInvalid `prisma.post.delete()` invocation:\n\n\n  The change you are trying to make would violate the required relation 'CommentToPost' between the `Comment` and `Post` models.",
```

Even though the `comments` table was altered for CASCADE deletes, it doesn't seem to work.
I had to do the following to get it working. I'm sure everyone would prefer not to do it this way though.

```typescript
t.nonNull.field('deletePost', {
  type: 'Boolean',
  args: {
    postId: nonNull(intArg())
  },
  resolve: async (_root, args, context) => {
    await context.prisma.comment.deleteMany({
      where: { postId: args.postId },
    })

    await context.prisma.post.delete({
      where: { id:  args.postId },
    })

    return true
  },
})
```

**Comment by ryands17** [comment, marked as answer]
Hey @awey95 👋 

Cascade deletes for now will only work if your relation is optional. There's an issue [here](https://github.com/prisma/prisma/issues/2057) that tracks this.

As a workaround, you can do the following with your schema:

```prisma
model Post {
  id       Int       @id @default(autoincrement())
  comments Comment[]
}

model Comment {
  id     Int   @id @default(autoincrement())
  post   Post? @relation(fields: [postId], references: [id])
  postId Int?
}
```

Add a `?` to the relation, run migrate over the changes and now cascade deletes will work fine.
Let me know if it works.
